codex/errors.nim

```diff
@@ -9,15 +9,22 @@
 
 import pkg/stew/results
 
+export results
+
 type
   CodexError* = object of CatchableError # base codex error
   CodexResult*[T] = Result[T, ref CodexError]
 
-template mapFailure*(
-    exp: untyped,
-    exc: typed = type CodexError
-): untyped =
+proc mapFailure*[T, V, E](
```

Comment:
I really don't think this should be a proc, we want this inlined as much as possible.

Reply:
Good point. It was giving compilation errors so I moved it to a proc to help figure them out. 

But I think a typed template would work as well. Especially now I know more about the source of the issues.

Reply:
Hmmm, so here's been the annoying bit. When `mapFailure` is a proc type everything compiles. 

However if I change `mapFailure` back to a template I get a compile error: 

```nim
/Users/jaremycreechley/projs/status/nim-codex/codex/manifest/types.nim(14, 11) Warning: imported and not used: 'questionable' [UnusedImport]
/Users/jaremycreechley/projs/status/nim-codex/codex/manifest/coders.nim(143, 40) template/generic instantiation of `mapFailure` from here
/Users/jaremycreechley/projs/status/nim-codex/codex/errors.nim(27, 6) Error: type mismatch: got <Result[cid.Cid, cid.CidError], proc (e`gensym76: CidError): ref CatchableError{.noSideEffect, gcsafe, locks: 0.} | proc (e`gensym70: CidError): ref CatchableError{.noSideEffect, gcsafe, locks: 0.}>
but expected one of:
func mapErr[T; E0: not void; E1: not void](self: Result[T, E0];
    f: proc (x: E0): E1): Result[T, E1]
  first type mismatch at position: 2
  required type for f: proc (x: E0: not void): E1: not void{.closure.}
  but expression 'convertToErr' is of type: None
func mapErr[T; E0: not void](self: Result[T, E0]; f: proc (x: E0)): Result[T,
    void]
  first type mismatch at position: 2
  required type for f: proc (x: E0: not void){.closure.}
  but expression 'convertToErr' is of type: None
2 other mismatching symbols have been suppressed; compile with --showAllMismatches:on to see them

expression: mapErr(init(Cid, blockBuf), convertToErr)
stack trace: (most recent call last)
/Users/jaremycreechley/.asdf/installs/nim/1.6.14/lib/system/nimscript.nim(429, 18)
/Users/jaremycreechley/projs/status/nim-codex/build.nims(41, 16) testTask
/Users/jaremycreechley/projs/status/nim-codex/build.nims(28, 8) testCodexTask
/Users/jaremycreechley/projs/status/nim-codex/build.nims(21, 15) test
/Users/jaremycreechley/projs/status/nim-codex/build.nims(18, 7) buildBinary
/Users/jaremycreechley/.asdf/installs/nim/1.6.14/lib/system/nimscript.nim(273, 7) exec
/Users/jaremycreechley/.asdf/installs/nim/1.6.14/lib/system/nimscript.nim(273, 7) Error: unhandled exception: FAILED: nim c --out:build/testCodex -d:codex_enable_proof_failures=true tests/testCodex.nim [OSError]
```

If I make the inner `convertToErr` into a template too I get compile errors elsewhere. These are similar errors that the original `mapFailure` template gave: 

```nim
/Users/jaremycreechley/projs/status/nim-codex/codex/manifest/types.nim(14, 11) Warning: imported and not used: 'questionable' [UnusedImport]
/Users/jaremycreechley/projs/status/nim-codex/codex/manifest/manifest.nim(103, 1) template/generic instantiation from here
/Users/jaremycreechley/projs/status/nim-codex/codex/errors.nim(27, 14) Error: convertToErr can raise an unlisted exception: Exception
stack trace: (most recent call last)
/Users/jaremycreechley/.asdf/installs/nim/1.6.14/lib/system/nimscript.nim(429, 18)
/Users/jaremycreechley/projs/status/nim-codex/build.nims(41, 16) testTask
/Users/jaremycreechley/projs/status/nim-codex/build.nims(28, 8) testCodexTask
/Users/jaremycreechley/projs/status/nim-codex/build.nims(21, 15) test
/Users/jaremycreechley/projs/status/nim-codex/build.nims(18, 7) buildBinary
/Users/jaremycreechley/.asdf/installs/nim/1.6.14/lib/system/nimscript.nim(273, 7) exec
/Users/jaremycreechley/.asdf/installs/nim/1.6.14/lib/system/nimscript.nim(273, 7) Error: unhandled exception: FAILED: nim c --out:build/testCodex -d:codex_enable_proof_failures=true tests/testCodex.nim [OSError]
```

Reply:
Phew, okay this version works: 

```nim
template mapFailure*[T, V, E](
    exp: Result[T, V],
    exc: typedesc[E],
): Result[T, ref CatchableError] =
  ## Convert `Result[T, E]` to `Result[E, ref CatchableError]`
  ##

  exp.mapErr(proc (e: V): ref CatchableError = (ref exc)(msg: $e))
```

That reminds me of the issue with `libp2p except setup`. I'd setup the inner `convertToErr` proc to try and resolve some other compile issues with type resolution. However, using a a named inner `convertToErr` proc/template blows up. However, an anonymous (typed) proc works! Confused yet? I am! 😂

Ok, so `libp2p` export issue is with the `asynctest` macros redefine a local `template setup` which has a different type and captures the `unittest.setup` template. That works until you import another `setup` name which seems to cause the whole template name resolution order to get weird.

However, Nim 1.6.14 includes fixes for all the defect / raises handling from the Chronos refactor. So... a bit of whack a mole.

Last commit works better.

Reply:
I think @emizzle 's changes to FuturesTracking actually helped fix up some of these issues. :)